File: java-accounting-app/src/Operations.java
final class Operations {
    enum Outcome { OK, INSUFFICIENT_FUNDS, EXCEEDS_LIMIT }

    record Result(Outcome outcome, long balanceCents) {}

    private Operations() {}

    static Result credit(long balance, long amount) {
        requireInputs(balance, amount);
        long next = Math.addExact(balance, amount);
        // L-01: refuse overflow instead of reproducing COBOL's discarded digit.
        return next > Money.MAX_CENTS
            ? new Result(Outcome.EXCEEDS_LIMIT, balance)
            : new Result(Outcome.OK, next);
    }

    static Result debit(long balance, long amount) {
        requireInputs(balance, amount);
        return amount > balance
            ? new Result(Outcome.INSUFFICIENT_FUNDS, balance)
            : new Result(Outcome.OK, balance - amount);
    }

    private static void requireInputs(long balance, long amount) {
        Money.requireBalance(balance);
        if (amount <= 0 || amount > Money.MAX_CENTS) {
            throw new IllegalArgumentException("amount must be positive and within the account limit");
        }
    }
}

File: dotnet-accounting-app/Operations.cs
namespace Accounting;

internal enum Outcome { Ok, InsufficientFunds, ExceedsLimit }

internal readonly record struct OperationResult(Outcome Outcome, long BalanceCents);

internal static class Operations
{
    internal static OperationResult Credit(long balance, long amount)
    {
        RequireInputs(balance, amount);
        var next = checked(balance + amount);
        // L-01: refuse overflow instead of reproducing COBOL's discarded digit.
        return next > Money.MaxCents
            ? new(Outcome.ExceedsLimit, balance)
            : new(Outcome.Ok, next);
    }

    internal static OperationResult Debit(long balance, long amount)
    {
        RequireInputs(balance, amount);
        return amount > balance
            ? new(Outcome.InsufficientFunds, balance)
            : new(Outcome.Ok, balance - amount);
    }

    private static void RequireInputs(long balance, long amount)
    {
        Money.RequireBalance(balance);
        if (amount <= 0 || amount > Money.MaxCents)
        {
            throw new ArgumentOutOfRangeException(nameof(amount), "Amount must be positive and within the account limit.");
        }
    }
}
